odl/diagnostics/operator.py

```diff
@@ -213,27 +214,40 @@ def adjoint(self):
         self._adjoint_definition()
         self._adjoint_of_adjoint()
 
-    def _derivative_convergence(self, step):
-        print('\nTesting derivative is linear approximation')
+    def _derivative_convergence(self):
+        name = 'Testing derivative is linear approximation'
 
-        with FailCounter("error = ||A(x+c*dx)-A(x)-c*A'(x)(dx)|| / "
-                         "|c| ||dx|| ||A||") as counter:
+        with FailCounter(name, "error = "
+                         "inf_c ||A(x+cp)-A(x)-A'(x)(cp)|| / c") as counter:
             for [name_x, x], [name_dx, dx] in samples(self.operator.domain,
                                                       self.operator.domain):
+                # Precompute some values
                 deriv = self.operator.derivative(x)
+                derivdx = deriv(dx)
                 opx = self.operator(x)
 
-                exact_step = self.operator(x + dx * step) - opx
-                expected_step = deriv(dx * step)
-                denom = step * dx.norm() * self.operator_norm
-                error = (0 if denom == 0
-                         else (exact_step - expected_step).norm() / denom)
+                c = 1e-4  # initial step
+                derivative_ok = False
 
-                if error > 0.00001:
-                    counter.fail('x={:15s} dx={:15s} c={}: error={:6.5f}'
-                                 ''.format(name_x, name_dx, step, error))
+                minerror = float('inf')
+                while c > 1e-14:
+                    exact_step = self.operator(x + dx * c) - opx
+                    expected_step = c * derivdx
+                    err = (exact_step - expected_step).norm() / c
+
+                    if err < 1e-4:
+                        derivative_ok = True
+                        break
+                    else:
+                        minerror = min(minerror, err)
 
-    def derivative(self, step=0.0001):
+                    c /= 2.0
```

Comment:
This will run 20 times in the worst case. Feels a bit much if we're doing that for all examples. How about dividing by 10?